drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2494 @@
+static int nvs_expire_ate(FAR struct nvs_fs *fs, uint32_t addr)
+{
+  uint8_t expired;
+  off_t offset;
+
+  expired = CONFIG_MTD_CONFIG_ERASEDVALUE;
+  expired = ~expired;
+  offset = fs->geo.erasesize * (addr >> ADDR_BLOCK_SHIFT);
+  offset += addr & ADDR_OFFS_MASK;
+
+  return flash_write(fs->mtd, offset + offsetof(struct nvs_ate, expired),
+                     &expired, sizeof(expired));
+}
+
+/****************************************************************************
+ * Name: nvs_gc
+ *
+ * Description:
+ *   garbage collection: the address ate_wra has been updated to the new
+ *   block that has just been started. The data to gc is in the block
+ *   after this new block.
+ *
+ ****************************************************************************/
+
+static int nvs_gc(FAR struct nvs_fs *fs)
+{
+  int rc;
+  struct nvs_ate close_ate;
+  struct nvs_ate gc_ate;
+  uint32_t sec_addr;
+  uint32_t gc_addr;
+  uint32_t gc_prev_addr;
+  uint32_t data_addr;
+  uint32_t stop_addr;
+
+  finfo("gc: before gc, ate_wra %" PRIx32 "\n", fs->ate_wra);
+
+  sec_addr = (fs->ate_wra & ADDR_BLOCK_MASK);
+  nvs_block_advance(fs, &sec_addr);
+  gc_addr = sec_addr + fs->geo.erasesize - sizeof(struct nvs_ate);
+
+  finfo("gc: set, sec_addr %" PRIx32 ", gc_addr %" PRIx32 "\n", sec_addr,
+        gc_addr);
+
+  /* if the block is not closed don't do gc */
+
+  rc = nvs_flash_ate_rd(fs, gc_addr, &close_ate);
+  if (rc < 0)
+    {
+      /* flash error */
+
+      return rc;
+    }
+
+  rc = nvs_ate_cmp_const(&close_ate, CONFIG_MTD_CONFIG_ERASEDVALUE);
+  if (!rc)
+    {
+      goto gc_done;
+    }
+
+  stop_addr = gc_addr - sizeof(struct nvs_ate);
+
+  if (nvs_close_ate_valid(fs, &close_ate))
+    {
+      gc_addr &= ADDR_BLOCK_MASK;
+      gc_addr += close_ate.offset;
+    }
+  else
+    {
+      rc = nvs_recover_last_ate(fs, &gc_addr);
+      if (rc)
+        {
+          return rc;
+        }
+    }
+
+  do
+    {
+      gc_prev_addr = gc_addr;
+      rc = nvs_prev_ate(fs, &gc_addr, &gc_ate);
+      if (rc)
+        {
+          return rc;
+        }
+
+      if (gc_ate.expired != CONFIG_MTD_CONFIG_ERASEDVALUE)
+        {
+          /* deleted or old ate, ignore it */
+
+          continue;
+        }
+
+      if (!nvs_ate_valid(fs, &gc_ate))
+        {
+          continue;
+        }
+
+      if (gc_ate.id != NVS_SPECIAL_ATE_ID)
+        {
+          /* copy needed */
+
+          finfo("Moving %" PRIu32 ", key_len %" PRIu16 ", len %" PRIu16 "\n",
+                gc_ate.id, gc_ate.key_len, gc_ate.len);
+
+          data_addr = (gc_prev_addr & ADDR_BLOCK_MASK);
+          data_addr += gc_ate.offset;
+
+          gc_ate.offset = (uint16_t)(fs->data_wra & ADDR_OFFS_MASK);
+          nvs_ate_crc8_update(&gc_ate);
+
+          rc = nvs_flash_block_move(fs, data_addr,
+                                    gc_ate.key_len + gc_ate.len);
+          if (rc)
+            {
+              return rc;
+            }
+
+          rc = nvs_flash_ate_wrt(fs, &gc_ate);
```

Comment:
how to guarantee the write never pass over the previous block?

Reply:
I don't get it , what is the meaning of  "pass over"?